tests/compile-fail/module_inception.rs

```diff
@@ -4,14 +4,27 @@
 
 mod foo {
     mod bar {
-        mod bar { //~ ERROR module has the same name as its containing module
+        pub mod bar { //~ ERROR item has the same name as its containing module
             mod foo {}
         }
         mod foo {}
     }
-    mod foo { //~ ERROR module has the same name as its containing module
+    pub mod foo { //~ ERROR item has the same name as its containing module
         mod bar {}
     }
 }
 
+mod cake {
+    mod cake {
+        // no error, since module is not public
```

Comment:
The goal of the lint is to catch a beginner's error of writing a redundant `mod foo` in `foo.rs`, so IMO it should not lint only on public modules.
Otherwise r+.